tests/binary_agreement.rs

```diff
@@ -1,108 +1,137 @@
-        .chain(once(rng.gen_range(30, 50)));
-    for size in sizes {
-        let num_faulty_nodes = (size - 1) / 3;
-        let num_good_nodes = size - num_faulty_nodes;
-        for &input in &[None, Some(false), Some(true)] {
-            info!(
-                "Test start: {} good nodes and {} faulty nodes, input: {:?}",
-                num_good_nodes, num_faulty_nodes, input
-            );
-            let adversary = |_| new_adversary(num_good_nodes, num_faulty_nodes);
-            let new_ba = |netinfo: Arc<NetworkInfo<NodeId>>| {
-                BinaryAgreement::new(netinfo, 0).expect("Binary Agreement instance")
-            };
-            let network = TestNetwork::new(num_good_nodes, num_faulty_nodes, adversary, new_ba);
-            test_binary_agreement(network, input);
-            info!(
-                "Test success: {} good nodes and {} faulty nodes, input: {:?}",
-                num_good_nodes, num_faulty_nodes, input
-            );
+impl VirtualNet<Algo> {
+    fn test_binary_agreement<R>(&mut self, input: Option<bool>, mut rng: R)
+    where
+        R: Rng + 'static,
+    {
+        let ids: Vec<NodeId> = self.nodes().map(|n| *n.id()).collect();
+        for id in ids {
+            let _ = self.send_input(id, input.unwrap_or_else(|| rng.gen::<bool>()));
         }
-    }
-}
 
-#[test]
-fn test_binary_agreement_random_silent() {
-    let new_adversary = |_: usize, _: usize| SilentAdversary::new(MessageScheduler::Random);
-    test_binary_agreement_different_sizes(new_adversary);
+        // Handle messages in random order until all nodes have output the proposed value.
+        while !self.nodes().all(|node| node.algorithm().terminated()) {
+            let _ = self.crank_expect();
+        }
+        // Verify that all instances output the same value.
+        let mut expected = input;
+        for node in self.nodes() {
+            if let Some(b) = expected {
+                assert!(once(&b).eq(node.outputs()));
+            } else {
+                assert_eq!(1, node.outputs().len());
+                expected = Some(node.outputs()[0]);
+            }
+        }
+        // TODO: As soon as observers are added to the test framework, compare the expected output
+        // against the output of observers.
+    }
 }
 
-#[test]
-fn test_binary_agreement_first_silent() {
-    let new_adversary = |_: usize, _: usize| SilentAdversary::new(MessageScheduler::First);
-    test_binary_agreement_different_sizes(new_adversary);
+/// Tests Binary Agreement on a given configuration with random inputs, with all `false` inputs and
+/// with all `true` inputs.
+#[cfg_attr(feature = "cargo-clippy", allow(needless_pass_by_value))]
+fn binary_agreement(cfg: TestConfig) {
+    let mut rng: TestRng = TestRng::from_seed(cfg.seed);
+    let size = cfg.dimension.size();
+    let num_faulty_nodes = cfg.dimension.faulty();
+    let num_good_nodes = size - num_faulty_nodes;
+    for &input in &[None, Some(false), Some(true)] {
```

Comment:
Should this be handled by proptest, too?

Reply:
OK, will do.